Excise commit (add Contact)

yutokyokutyo opened the pull request:
@tacahilo @kitak @gs3 @keokent

お疲れ様です。
Gitで詰まってしまい、時間がかかってしまいました。
(くろさんに助言をいただき解決しました！)

Sample_app にContact ページを追加しました。
MasterにMargeをしたいのですが、レビューをいただけますでしょうか。
よろしくお願いいたします。

Greenを確認しました。
![2014-07-08 18 00 02](https://cloud.githubusercontent.com/assets/7356977/3507442/641e6b20-067e-11e4-9946-52938d9400de.png)
http://192.168.25.37/gyazo/data/2e183393b89fcf586eebd0960f8fd7d3.png

ローカルサーバでの表示を確認しました。
![2014-07-08 18 00 11](https://cloud.githubusercontent.com/assets/7356977/3507456/7992f444-067e-11e4-9b21-6bc2c8f36e49.png)
http://192.168.25.37/gyazo/data/98099f15d1548cbcb3dcbfa3238c8585.png

yutokyokutyo committed:
Excise commit (add Contact)

kitak commented on app/controllers/static_pages_controller.rb:
```diff
@@ -7,4 +7,7 @@ def help
   def about
   end
+  
+  def Contact
```
Rubyだと名前の先頭が大文字だと、定数、モジュールを連想するので、メソッド名の先頭の文字は小文字がいいと思います。

このコーディング規約(http://shugo.net/ruby-codeconv/codeconv.html の「メソッド名」)にも、書かれてるよ。
コーディング規約は、cookpadの公開しているもの、githubの公開しているもの、と色々あるので時間のあるときに眺めてみよう。

keokent commented on config/routes.rb:
```diff
@@ -2,6 +2,7 @@
   get "static_pages/home"
   get "static_pages/help"
   get "static_pages/about"
+  get "static_pages/Contact"
```
`/static_pages/Contact` のように大文字始まりってあまりみないかなぁ

hfm replied:
home, help, aboutとあって、Contactだと、ルールが統一されてないから、合わせたほうがいいんじゃないかな。

hfm commented:
とりあえずテスト通ったー。

``` console
usr0600296@okkun [~/src/github.com/yutokyokutyo/sample_app] [6-1] (master[-] @0) [perl v5.18.2] [ruby 2.1.1p76] [python 2.7.7]
$ bin/rake spec
/Users/usr0600296/.rbenv/versions/2.1.1/bin/ruby -S rspec ./spec/requests/static_pages_spec.rb
......

Finished in 0.58809 seconds
6 examples, 0 failures

Randomized with seed 10940
```

hfm commented on spec/requests/static_pages_spec.rb:
```diff
@@ -40,4 +40,17 @@
     end
   end
+    
+    describe "Contact" do
```
細かいけどインデントが右にズレてしまってるね。

yutokyokutyo commented:
レビューありがとうございます！
修正してみます！

yutokyokutyo committed:
Exercises commit (add contact fix)

yutokyokutyo commented:
## 3.5 Exercises 1

> http://www.railstutorial.org/book/static_pages#sec-static_pages_exercises

コーディング規約とっても勉強になります！
時間が空いているときに、ちょこちょこ読み進めていこうと思います。

昨日ご指摘いただいた点について修正を加えました。
レビューをお願いいたします！

### 動作確認

![2014-07-09 14 33 56](https://cloud.githubusercontent.com/assets/7356977/3520360/90262bc6-072e-11e4-8466-11c7256b2305.png)
- Greenを確認しました。

![2014-07-09 14 35 59](https://cloud.githubusercontent.com/assets/7356977/3520361/953bb0f4-072e-11e4-868e-0546917478ec.png)
- ローカルサーバでの表示を確認しました。

@tacahilo @kitak @gs3 @keokent

yutokyokutyo commented:
[![LGTM](http://www.lgtm.in/p/KZdSzolaq)](http://www.lgtm.in/i/KZdSzolaq)

(どうしてもLGTMを使ってみたくて使ってみた。)

### 併せて、プルリクの仕方についても何かレビューをいただけたら幸いです。

hfm commented:
LGTMを自画自賛に使うのを初めてみた笑

keokent commented:
良いと思う！ :hamster:

hfm commented:
落ちた

``` console
$ bin/rake spec
/Users/usr0600296/.rbenv/versions/2.1.1/bin/ruby -S rspec ./spec/requests/static_pages_spec.rb
....FF..

Failures:

  1) Static pages Contact should have the title 'Contact'
     Failure/Error: visit '/static_pages/contact'
     ActionView::MissingTemplate:
       Missing template static_pages/contact, application/contact with {:locale=>[:en], :formats=>[:html], :handlers=>[:erb, :builder, :raw, :ruby, :jbuilder, :coffee]}. Searched in:
         * "/Users/usr0600296/src/github.com/yutokyokutyo/sample_app/app/views"
     # ./spec/requests/static_pages_spec.rb:52:in `block (3 levels) in <top (required)>'

  2) Static pages Contact should have the content 'Contact'
     Failure/Error: visit '/static_pages/contact'
     ActionView::MissingTemplate:
       Missing template static_pages/contact, application/contact with {:locale=>[:en], :formats=>[:html], :handlers=>[:erb, :builder, :raw, :ruby, :jbuilder, :coffee]}. Searched in:
         * "/Users/usr0600296/src/github.com/yutokyokutyo/sample_app/app/views"
     # ./spec/requests/static_pages_spec.rb:47:in `block (3 levels) in <top (required)>'

Finished in 0.10969 seconds
8 examples, 2 failures

Failed examples:

rspec ./spec/requests/static_pages_spec.rb:51 # Static pages Contact should have the title 'Contact'
rspec ./spec/requests/static_pages_spec.rb:46 # Static pages Contact should have the content 'Contact'

Randomized with seed 17889

rake aborted!
/Users/usr0600296/.rbenv/versions/2.1.1/bin/ruby -S rspec ./spec/requests/static_pages_spec.rb failed
/Users/usr0600296/src/github.com/yutokyokutyo/sample_app/vendor/ruby/2.1.0/gems/rspec-core-2.13.1/lib/rspec/core/rake_task.rb:156:in `run_task'
/Users/usr0600296/src/github.com/yutokyokutyo/sample_app/vendor/ruby/2.1.0/gems/rspec-core-2.13.1/lib/rspec/core/rake_task.rb:124:in `block (2 levels) in initialize'
/Users/usr0600296/src/github.com/yutokyokutyo/sample_app/vendor/ruby/2.1.0/gems/rspec-core-2.13.1/lib/rspec/core/rake_task.rb:122:in `block in initialize'
Tasks: TOP => spec
(See full trace by running task with --trace)
```

hfm commented:
多分ここのファイル名が更新されてないっぽい 

![478abbc70d143c4d9761509365209e89](https://cloud.githubusercontent.com/assets/1348959/3522128/ab2a79a8-0748-11e4-8065-90ae6d071b7d.png)

keokent commented:
> 落ちた

おっ　テストも回さないでレビューをしてしまった。。。

keokent commented:
> ActionView::MissingTemplate

アクション名が変わったらviewのファイルの名前も変更しないといけないね

yutokyokutyo commented:
@tacahilo @keokent 

おおお。
 `bin/rake spec` でもテストを走らせることができるのですね。
ブラウザにContactページを表示させることができたので、自分の中ではこれで大丈夫だろうと思っていました。
危ない危ない。
修正します！

いつも`bundle exec rspec spec/requests/static_pages_spec.rb` でテストをしておりましたが、
これでは不十分なのですね。
 `bin/rake spec` との違いは何なのでしょうか。
調べてみて分からなかったら後でお聞きしようと思います！

yutokyokutyo committed:
Exercises commit (add contact fix2)
	renamed:    app/views/static_pages/Contact.html.erb -> app/views/static_pages/contact.html.erb

yutokyokutyo commented:
@tacahilo @kitak @gs3 @keokent

> renamed:    app/views/static_pages/Contact.html.erb -> app/views/static_pages/contact.html.erb

以上のように修正を施しました。
レビューをお願いいたします。

### 動作確認

![2014-07-09 16 27 53](https://cloud.githubusercontent.com/assets/7356977/3521068/c22f4c22-073a-11e4-98e8-75d907748209.png)
- `bundle exec rspec spec/requests/static_pages_spec.rb`  により、Greenを確認しました

![2014-07-09 16 30 01](https://cloud.githubusercontent.com/assets/7356977/3521077/df232f4c-073a-11e4-87bb-7d44830a351b.png)
- ローカルサーバでの表示を確認しました。

---

先程も上のようにテストをして問題ないと思ったのですが、結果としてviewのファイルの名前を変更していないことがわかりました。
これでは `bundle exec rspec` でのテストの意味が薄い気がするのですが、実際の業務等では `bin/rake spec` のように、他のテストもかけたりしているのでしょうか。。

hfm commented:
:sparkles: 通ったー :sparkles: 

``` console
$ bin/rake spec
/Users/usr0600296/.rbenv/versions/2.1.1/bin/ruby -S rspec ./spec/requests/static_pages_spec.rb
........

Finished in 0.12602 seconds
8 examples, 0 failures

Randomized with seed 31152
```

hfm commented:
> これでは bundle exec rspec でのテストの意味が薄い気がするのですが、実際の業務等では bin/rake spec のように、他のテストもかけたりしているのでしょうか。。

ここは普段からrubyを書いている @keokent @gs3 @kitak に期待。

yutokyokutyo commented:
実際の業務だとリモートの部分では、Githubにpushしてから Jenkinsが自動でテストしてたなあというのは覚えているのですが、自分のローカルでのテストの手段はどういったものがあるのか知りたい・・です。

ravelll commented:
@yutokyokutyo 

> https://github.com/yutokyokutyo/sample_app/pull/1#issuecomment-48435358
> いつもbundle exec rspec spec/requests/static_pages_spec.rb でテストをしておりましたが、これでは不十分なのですね。
> bin/rake spec との違いは何なのでしょうか。

まず、`bin/rake spec` と `bundle exec rspec` とで、テスト内容に差はありません。
ですが、`bundle exec rspec spec/requests/static_pages_spec.rb` としたときには `spec/requests/static_pages_spec.rb` の内容のみについて、
`bin/rake spec` では `spec/` 以下の全てのファイルについてのテストとなります。

ので、
- 特定の部分だけテストしたいときはその部分についての spec ファイル名を入れてテスト
- アプリケーション全体が壊れていないか見たいときはファイル名を入れず全体をテスト

のように使い分けられます！

kitak commented:
> 他のテストもかけたりしているのでしょうか

(あくまで僕の場合ですが)
ファイルを変更するたびにすべてのテストを走らせると時間がかかってしまうので、ファイルの修正をおこなった場合は修正と関連のあるテストを狙って走らせるようにしています。
作業が一段落してpushするタイミングで、どこか壊れていないか確かめるためにすべてのテストを走らせています。

kitak commented:
みてて思ったんだけど、なんで、https://github.com/yutokyokutyo/sample_app/pull/1#issuecomment-48433335 のタイミングでテストが全部通ったんだろう...

`ActionView::MissingTemplate`でテストが落ちそうな気がするんだけどなぁ

kitak commented:
あ、こういうことか。失礼しました。

```
15:33 okkun: macだと、っていうの良くないっすね。正確な名前忘れましたけど、大文字と小文字の区別つけないファイルシステムの設定になってしまってる可能性が。
15:33 udzura: OS X のデフォルトですねー（つらい
```

yutokyokutyo commented:
@tacahilo @gs3 @kitak 

なるほどお！！
Unit test と Integration test についてkitakさんのテスト開発講義で教えていただいたのを思い出しました！
つまり、

`bundle exec rspec spec/requests/static_pages_spec.rb` 　：　Unit test
`bin/rake spec`　：　Integration test

のように分類することができ、Testをする目的や時間とのバランスを考えながらTestの種類を変えていくということですね！
実際の業務で、Testをどのように使い分けているかという部分も大変参考になりました！
ありがとうございました！

お手間をとらせてしまい申し訳ありませんでした。
これから少しづつ自分で解決できるようになっていきたいと思います。

keokent commented:
@yutokyokutyo 
MTGでレビューが遅くなっていしまいました。

> 他のテストもかけたりしているのでしょうか

ぼくも @kitak  とだいたい同じです。手元では修正と関連するところのテストだけをローカルで回して、`bundle exec rspec spec/` のようなに全部のテストを行うのは Push したときに Jenkins が自動で走らせてくれるので、そこで結果を見るようにしています。

kitak commented:
@yutokyokutyo 

> bundle exec rspec spec/requests/static_pages_spec.rb 　：　Unit test
> bin/rake spec　：　Integration test

というのは、ちょっとちがうぞ！

講義でいってた Unit test は、モデルとかコントローラのようなひとつの部品に着目したテスト(例えば、`User`モデルのテスト)のことで、Integration test は複数の部品にまたがったテスト(例えば、今回の章のようなページにアクセスして表示内容を確認するようなテスト)のことです。

まだTutorial でIntegration testしか登場していないから誤解をしちゃったかもしれないけど、 今後のChapterでモデルのテスト(Unit test)を書いたりもするので、そのときにイメージがつかめると思うよ！

keokent commented:
今回行った `bundle exec rspec spec/requests/static_pages_spec.rb` は 数あるテストのうち、ファイル名を指定して、static_pages の Integration test のみを実行した と言った感じです。

yutokyokutyo committed:
【3.5 Exercises 2 Commit 】RSpec let function

ravelll commented:
@yutokyokutyo 

おや！たしかに、
- ある実装（関数）についての細かなテスト＝単体（Unit）テスト
- ある機能を実行したときの流れ全体のテスト＝結合（Integration）テスト

のような話はあったけれど、それらはテストの書き方の話で、テスト実行の仕方の話とはまた違う話なのです :smiley: 
例えばある機能 X について、内部ではそれを実行するために処理をメソッド `def hogeA`, `def hogeB`, `def hogeC` の3つに役割分担していたとして、
- 単体テストは `describe "hogeA"` のような、内部に作った細かい役割についてのテストのこと
  - 主にコントローラーとモデルが対象
- 結合テストは `describe "Kinou X"` のような、ある機能を使ったときどうなる、なテストのこと
  - ビュー、コントローラー、モデルを横断するテスト

な感じです！（わからなかったらごめん！）

yutokyokutyo commented:
@tacahilo @gs3 @kitak

勘違いしてしまったようですね。
お三方のコメントで大分理解できたかと思います！
これからUnit testについてもRails Tutorial に出てくるということなので、そこでさらに知見を深めていきたいです！
ありがとうございました！:sushi:

yutokyokutyo commented:
Merge します！

keokent commented:
おつかれさま！ :hamster:

hfm commented:
おつおつ。知見のたまるPRでした :bow:

yutokyokutyo commented:
ryomaさんに今日の夕会で共有したいと思います。
ありがとうございました！！
